tokio/src/task/preemption.rs

```diff
@@ -0,0 +1,115 @@
+//! with the executor to avoid starvation.
+//!
+//! Consider a future like this one:
+//!
+//! ```
+//! # use tokio::stream::StreamExt;
+//! async fn drop_all<I: Stream>(input: I) {
+//!     while let Some(_) = input.next().await {}
+//! }
+//! ```
+//!
+//! It may look harmless, but consider what happens under heavy load if the input stream is
+//! _always_ ready. If we spawn `drop_all`, the task will never yield, and will starve other tasks
+//! and resources on the same executor. With opt-in preemption, this problem is alleviated:
+//!
+//! ```
+//! # use tokio::stream::StreamExt;
+//! async fn drop_all<I: Stream>(input: I) {
+//!     while let Some(_) = input.next().await {
+//!         tokio::preempt_check!();
+//!     }
+//! }
+//! ```
+//!
+//! The call to [`preempt_check!`] will coordinate with the executor to make sure that every so
+//! often control is yielded back to the executor so it can run other tasks.
+//!
+//! # Placing preemption points
+//!
+//! Preemption points should be placed _after_ at least some work has been done. If they are not, a
+//! future sufficiently deep in the task hierarchy may end up _never_ getting to run because of the
+//! number of preemption points that inevitably appear before it is reached.
+//!
+//!   [`preempt_check!`]: ../../macro.check.html
+
+use std::cell::Cell;
+use std::task::{Context, Poll};
+
+/// Constant used to determine how much "work" a task is allowed to do without yielding.
+///
+/// The value itself is chosen somewhat arbitrarily. It needs to be high enough to amortize wakeup
+/// and scheduling costs, but low enough that we do not starve other tasks for too long. The value
+/// also needs to be high enough that particularly deep tasks are able to do at least some useful
+/// work at all.
+///
+/// Note that as more preemption points are added in the ecosystem, this value will probably also
+/// have to be raised.
+const BUDGET: usize = 128;
+
+thread_local! {
+    static HITS: Cell<usize> = Cell::new(0);
+}
+
+/// Mark that the top-level task yielded, and that the preemption budget should be reset.
+pub(crate) fn yielded() {
+    HITS.with(|hits| {
+        hits.set(BUDGET);
+    });
+}
+
+/// Returns `Poll::Pending` if the current task has exceeded its preemption budget and should yield.
+pub fn poll(cx: &mut Context<'_>) -> Poll<()> {
+    HITS.with(|hits| {
+        let n = hits.get();
+        if n == 0 {
+            cx.waker().wake_by_ref();
+            Poll::Pending
+        } else {
+            hits.set(n.saturating_sub(1));
+            Poll::Ready(())
+        }
+    })
+}
+
+#[doc(hidden)]
+pub async fn maybe_yield() {
+    use crate::future::poll_fn;
+    poll_fn(|cx| poll(cx)).await;
+}
+
+/// Yield if this `async` block's task has exceeded its preemption budget.
+#[macro_export]
+macro_rules! preempt_check {
+    () => {
+        $crate::task::preemption::maybe_yield().await
+    };
+}
```

Comment:
I'm not sure if it's worth putting this in a macro _just_ to avoid having to type `.await`...if we just publicly exposed the `maybe_yield` function, users could import it from `tokio::task::preemption` rather than from the root...